Investigate machine type architecture performance (#75)
ONSdigital/eq-survey-runner-benchmark: doc/performance-investigations/0012-increase-machine-type-core-count/summary.md

# Investigate Machine Type Architecture Performance

EQ Runner currently uses a machine type of `custom-4-4096` (4 vCPU cores, 4096MB of RAM). This investigation looked at the impact of using larger instances with more pods on the performance of the runner application.

## Benchmark profile

| Option                 | Value                        |
| ---------------------- | ---------------------------- |
| Requests file          | census_household_gb_eng.json |
| Run time               | 20m                          |
| User wait time minimum | 1                            |
| User wait time maximum | 2                            |

> | Clients                | 256                          |

Reviewer: It might be clear if that was `Clients per core` which was 16 just to make it clear each pod wasn't run with 256.

Author: good point 👍🏻

# Investigate Machine Type Architecture Performance

> EQ Runner currently uses a machine type of `custom-4-4096` (4 vCPU cores, 4096MB of RAM). This investigation looked at the impact of machine type on the performance of the runner application, focusing on the number of CPU cores as Runner is primarily a CPU bound application with limited network IO to Datastore and Redis.

Reviewer 2: I'm not sure about limited network IO, the calls to Datastore are up to 50% of the request time.

Author: How about `EQ Runner currently uses a machine type of `custom-4-4096` (4 vCPU cores, 4096MB of RAM). This investigation looked at the impact of machine type on the performance of the runner application, focusing on the number of CPU cores provisioned` ?

Reviewer 2: After re-reading this, I think it should be clearer about what we're testing. Something like: 
> This investigation looked at the impact of machine type on the performance of the runner application, specifically the impact of using larger instances with more pods.

> The investigation aims to determine whether latencies as measured by the Locsut client improve when using larger machines but utilizing roughly the same percentage of CPU.

Reviewer 2: ```suggestion
The investigation aims to determine whether latencies as measured Locust improve when using larger machines but utilising roughly the same percentage of CPU.
```

## Benchmark profile

| Option                 | Value                        |
| ---------------------- | ---------------------------- |
| Requests file          | census_household_gb_eng.json |
| Run time               | 20m                          |
| User wait time minimum | 1                            |
| User wait time maximum | 2                            |
| Clients                | 256                          |
| Number of workers      | 7                            |
| Number of threads      | 7                            |

> This test replicates the current settings for the daily test. Note that the baseline test ran with 64 clients as per the daily test.

Reviewer 2: What baseline test?

Author: I can remove this from here, as it references the comparison against 4 cores and 1 pod in the c type test.

## Results

| Cores | Pods | Nodes | GET Average | POST Average | All Request Average | Total Requests |
| ----- | ---- | ----- | ----------- | ------------ | ------------------- | -------------- |
| 4     | 4    | 4     | 572         | 609          | 590                 | 403,890        |
| 8     | 4    | 2     | 542         | 579          | 560                 | 404,535        |

> | 8     | 4    | 2     | 591         | 601          | 596                 | 402,884        |

Reviewer 2: What is different about these tests?

Author: Nothing, they were run twice so I've included both

Reviewer 2: It makes it inconsistent, why was 8 cores run twice and not the others?

Author: Purely coincidental - the environment was up and ran at it's triggered time. Yes it's inconsistent, but it highlights how the figures are variable enough that we can glean very little info from the results.

Reviewer 2: Maybe averaging them would be better then? If we don't think the results are reliable, then that brings into question the validity of the whole test.

Author: I think averaging just hides the variability. In terms of validity, I don't think the validity is a question, more that due to the variability of the tests, any improvement within the bounds of the variation can't be distinguished from the noise we see.

Author: I'm happy to remove the second run from the table and leave the data in the outputs

| 16    | 4    | 1     | 620         | 651          | 636                 | 138,195        |

> | 16    | 4    | 1     | 579         | 610          | 594                 | 139,156        |

Reviewer 2: What is different about these tests?

Author: Copy paste error. The second 16 4 1 test Should have been 16 5 1

Reviewer 2: The total requests still looks incorrect?

Author: Are you viewing an old version of the file? Should be 379,246 and 394,487

Reviewer 2: Looks like it 🤦

> | 16    | 5    | 1     | 557         | 598          | 577                 | 379,246        |

Reviewer 2: The numbers for 4 pods are incredibly low compared with 5 pods, do we know why?

Author: Copy and paste error from the google sheet tracking this. I'll update

In addition, a test was run against a node provisioned using a C type machine. C type machines have better per-core clock speeds, making them ideal for single-threaded applications. These tests were run with 64 clients.

| Architecture | Cores | Pods | Nodes | GET Average | POST Average | All Request Average | Total Requests |
| ------------ | ----- | ---- | ----- | ----------- | ------------ | ------------------- | -------------- |
| Default      | 4     | 1    | 1     | 239         | 288          | 263                 | 128,808        |
| C Type       | 4     | 1    | 1     | 136         | 113          | 125                 | 139,227        |

## Observations

> - There is no significant difference in performance when using larger numbers of cores.

Reviewer 2: ```suggestion
- There is no significant difference in performance when using machines with more cores.
```

> - There is a significant performance gain per core when using C type machines.

Reviewer 2: ```suggestion
- There is a significant performance gain when using C type machines.
```

## Decision

- Continue to run using the 4 core machines currently provisioned.

> - If the number of nodes that are required becomes the limiting factor in horizontal scaling, then larger machine types could be utilized.

Reviewer 2: ```suggestion
- If the number of nodes that are required becomes the limiting factor in horizontal scaling, then larger machine types could be utilised.
```

> - Further investigation of C type machines should be carried.

Reviewer 2: Carried where? 😛

Author: Over the hills and far away? 😄

Author: Done in c7bcc07

> - A further test of larger machines should be carried out, whereby the number of Gunicorn workers and threads are scaled relative to the available cores.

Reviewer 2: I think it would be good to be clear here that it would be one runner pod per node.
```suggestion
- A further test of larger machines should be carried out, using a single runner pod per node and scaling the number of Gunicorn workers and threads relative to the available cores.
```
